### PR TITLE
Add compiler wrapper behavior

### PR DESCRIPTION
**Pull request by nicovank**
One atomic PR to add compiler wrapper behavior.

This mode is more robust because filenames can properly be resolved from the same working directory as the compiler.

### Issues

 -  A bit more overhead for successful compilation from Python startup time.
 -  Configuration (such as CMake) can be inefficient if checking for things that fail since we will do an API call. Maybe we should add an environment variable check to allow disabling CWhy during configuration.

### How it works

A short script is generated in a temporary file. This allows options to be serialized along with the script and stay consistent across runs. I experimented with a couple other ways to do this but I think this is the best one.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,31 +17,56 @@ Explains and suggests fixes for C/C++/Rust compiler error messages.
...
+
+Just pipe your compiler's output to `cwhy`.
+
+```bash
+% clang++ -g mycode.cpp |& cwhy
 ```
 
+### Options
+
+These options can be retrieved with `cwhy --help`.
```
Comment (emeryberger) [inline README.md:56]: change "retrieved" to "displayed".

File: README.md [hunk 2]
```diff
@@ -16,31 +17,56 @@ Explains and suggests fixes for C/C++/Rust compiler error messages.
...
+
+ -  `--llm`: pick a specific OpenAI LLM. CWhy is tested with `gpt-3.5-turbo` and `gpt-4`.
```
Comment (emeryberger) [inline README.md:58]: "Cwhy has been tested with"

File: src/cwhy/__main__.py [hunk 3]
```diff
@@ -37,42 +66,46 @@ def main():
...
-def extract_sources(args):
-    cwhy.evaluate_prompt(
-        args,
-        cwhy.extract_sources_prompt(sys.stdin.read()),
-        wrap=False,
-    )
+    if args["subcommand"] == "wrapper":
+        with tempfile.NamedTemporaryFile(mode="w", delete=False) as f:
+            f.write(wrapper(args))
+        os.chmod(f.name, 0o755)
```
Comment (emeryberger) [inline src/cwhy/__main__.py:95]: Add a comment here to explain the octal string

File: src/cwhy/cwhy.py [hunk 4]
```diff
@@ -226,11 +217,6 @@ def __init__(self, diagnostic):
 
 def extract_sources_prompt(diagnostic):
     ctx = extract_sources_context(diagnostic)
```
Comment (emeryberger) [inline src/cwhy/cwhy.py:219]: What happens now if stdin is empty? (Someone just types `cwhy` or the compiler outputs no errors...)

File: src/cwhy/__main__.py [hunk 5]
```diff
@@ -2,11 +2,40 @@
...
 import argparse
 import importlib.metadata
+import os
 import sys
+import tempfile
 
 from . import cwhy
 
 
+def wrapper(args):
```
Comment (emeryberger) [inline src/cwhy/__main__.py:12]: Ugh, a giant string. Any reason for this? I'd rather it be code.

Comment (emeryberger) [reply src/cwhy/__main__.py:12]: You could even introspect on the function body using `inspect`.

**Review by emeryberger — CHANGES_REQUESTED** [review]

File: src/cwhy/cwhy.py (hunk 4 above)
Comment (nicovank) [reply src/cwhy/cwhy.py:219]: I just factored out all the `stdin == ""` checks into main: `if stdin`. Behavior is unchanged.

File: src/cwhy/__main__.py [hunk 6]
```diff
@@ -37,42 +45,47 @@ def main():
...
+        with tempfile.NamedTemporaryFile(mode="w", delete=False) as f:
+            f.write(wrapper(args))
+        # NamedTemporaryFiles are not executable by default. Set its mode to 755 here with an octal literal.
+        os.chmod(f.name, 0o755)
+        print(f.name)
+    elif args["version"]:
+        print(f"cwhy version {importlib.metadata.metadata('cwhy')['Version']}")
+    else:
+        stdin = sys.stdin.read()
+        if stdin:
```
Comment (nicovank) [inline src/cwhy/__main__.py:81]: Early exit if `stdin` is empty is now here.

**Review by emeryberger — APPROVED** [review]

**Pull request merged**